internal/upload/oci/upload.go

```diff
@@ -0,0 +1,234 @@
+package oci
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/oracle/oci-go-sdk/v51/common"
+	"github.com/oracle/oci-go-sdk/v51/core"
+	"github.com/oracle/oci-go-sdk/v51/identity"
+	"github.com/oracle/oci-go-sdk/v51/objectstorage"
+	"github.com/oracle/oci-go-sdk/v51/objectstorage/transfer"
+	"github.com/oracle/oci-go-sdk/v51/workrequests"
+)
+
+type Uploader interface {
+	Upload(name string, bucketName string, namespace string, file *os.File, user, compartment string) (string, error)
+	GetBucket(bucketName string, namespace string) (objectstorage.Bucket, error)
+}
+
+type ImageCreator interface {
+	create(imageName, bucketName, namespace, compartmentID string) (string, error)
+}
+
+type Client struct {
+	Uploader
+	ImageCreator
+	ociClient
+}
+
+// Upload uploads a file into an objectName under the bucketName in the namespace.
+func (c Client) Upload(objectName string, bucketName string, namespace string, file *os.File, userID, compartmentID, imageName string) (string, error) {
+	uploadManager := transfer.NewUploadManager()
+	getBucket, err := c.GetBucket(bucketName, namespace)
+	if err != nil {
+		return "", fmt.Errorf("failed to find a bucket with name %s under namespace %s: %w", bucketName, namespace, err)
+	}
+	req := transfer.UploadFileRequest{
+		UploadRequest: transfer.UploadRequest{
+			NamespaceName: getBucket.Namespace,
+			BucketName:    getBucket.Name,
+			ObjectName:    common.String(objectName),
+			CallBack: func(multiPartUploadPart transfer.MultiPartUploadPart) {
+				if multiPartUploadPart.Err != nil {
+					log.Printf("upload failure: %s\n", err)
+				}
+				log.Printf("multipart upload stats: parts %d, total-parts %d\n",
+					multiPartUploadPart.PartNum,
+					multiPartUploadPart.TotalParts)
+			},
+			ObjectStorageClient: &c.storageClient,
+		},
+		FilePath: file.Name(),
+	}
+
+	ctx := context.Background()
+	resp, err := uploadManager.UploadFile(ctx, req)
+
+	if err != nil {
+		if resp.IsResumable() {
+			resp, err = uploadManager.ResumeUploadFile(ctx, *resp.MultipartUploadResponse.UploadID)
+			if err != nil {
+				return "", err
+			}
+		} else {
+			return "", fmt.Errorf("failed to upload the file to object %s:  %w", objectName, err)
+		}
+
+	}
+	// clean up the object even if we fail
+	defer c.deleteObjectFromBucket(objectName, bucketName, namespace)
+
+	imageID, err := c.create(objectName, bucketName, namespace, compartmentID, imageName)
+	if err != nil {
+		return "", fmt.Errorf("failed to create a custom image using object '%s' bucket '%s' in namespace '%s': %w",
+			objectName,
+			bucketName,
+			namespace,
+			err)
+	}
+	err = c.GrantImageAccess(imageID, userID, compartmentID)
```

Comment:
I wonder if it would be better to expose `Create()` as public and use both `Create()` and `GrantImageAccess()` separately from `Upload()`, like we do with AWS.  It will make the external uploading and registering more granular (which could be better or worse), but it also makes it possible to have object uploads to OCI in the future, like we have with S3.

Reply:
First, I read my code again and decide to rename create -> createImage.
For having more options to upload, I think that instead of exposing the low level methods, we can
make the oci.Client use a different Uploader implementation. And we can build it using 2 differnt constructor functions:
```
NewCustomImageUploader() (Client, error)

NewBucketImageUploader() (Client, error)
```

Reply:
That would also work.  So the `BucketImageUploader` would just upload an image to a bucket as a blob, and the `CustomImageUploader` would do the registration as well?
They'd both be reusing the bucket upload.

I think that it might obfuscate what's going on when someone's reading the caller code in the worker.  I'm more in favour of sort of self-documenting things.  The way we have it for AWS, which I think makes the flow clearer is:
```go
			_, err = a.Upload(path.Join(outputDirectory, exportPath, options.Filename), options.Bucket, key)
			if err != nil {
				...
			}
			ami, err := a.Register(args.Targets[0].ImageName, options.Bucket, key, options.ShareWithAccounts, common.CurrentArch())
			if err != nil {
				...
			}
			if ami == nil {
				...
			}
```

and I think that makes it very easy to trace the whole thing (though admittedly, I cleaned up the error handling code).